### PR TITLE
Changed index search by binary search for sparse glyphs fonts.

### PR DESCRIPTION
**Pull request by beibean**
Pull request to use binary search as commented in #915.

Note that function **lv_font_codeCompare** could be moved to lv_math for example if any other module needs to use binary search and same kind of comparer.

*Tested within device with MK22FN1M0AXXX12*

### CODE REVIEW

**Comment by kisvegabor** [comment]
Hi,

Thank you!

Can we sure that `bsearch` is available in every embedded compiler?

**Comment by beibean** [comment]
@kisvegabor It should, since it is standard C function.

**Comment by beibean** [comment]
I just updated lv_font.c fixing a mistake.

bsearch was using sizeof(uint32_t*) instead of sizeof(uint32_t). I did not realize this yesterday since the micro I am using is 32-bits arquitecture.

Today I am using VS simulation which is 64-bits arquitecture, that's why I saw this problem.

**Comment by embeddedt** [comment]
The only problem I see here is that `bsearch` seems like a more "complex" C standard library function.

Many times I think people wind up implementing/borrowing subsets of the standard library that they need rather than using an existing one like newlib, for a few reasons:

1. Full libraries like newlib often require stubs for file support, heap, etc.
2. It allows more flexibility over the implementation and can make debugging easier.

I doubt most people who implemented functions like `strcpy` themselves would have bothered to implement `bsearch`.

Can we make it so that this is selectable with a configuration option?

**Comment by kisvegabor** [comment]
@embeddedt 
> Can we make it so that this is selectable with a configuration option?

Do you mean to implement `bsearch` and add a config to use the custom implementation or the library version? If yes, I think we should use only the custom implementation. I see no reason to use the library version once it is implemented here. Maybe a minor performance gain?

**Comment by beibean** [comment]
@embeddedt
> The only problem I see here is that bsearch seems like a more "complex" C standard library function.

I don't think bsearch is that "complex", you can see a valid implementation [here](https://github.com/torvalds/linux/blob/master/lib/bsearch.c).

**Comment by kisvegabor** [comment]
@beibean If it's that simple then I'd like to have this function in LittlevGL to avoid possible issues in the future. It can be in `lv_math.c` as `lv_bsearch()`.

**Comment by embeddedt** [comment]
@kisvegabor
> It can be in `lv_math.c` as `lv_bsearch()`

No need for a configuration option then. I just wanted it because I wasn't sure about the availability of `bsearch` on everyone's C library. If we implement it ourselves then that is not a problem.

**Comment by beibean** [comment]
@kisvegabor Added and tested lv_bsearch()

File: lv_misc/lv_math.c [hunk 1]
```diff
@@ -159,6 +159,49 @@ int32_t lv_bezier3(uint32_t t, int32_t u0, int32_t u1, int32_t u2, int32_t u3)
...
+ * the key and elements in the array are of the same type, you can use
+ * the same comparison function for both sort() and bsearch().
+ *
+ * @param key  pointer to item being searched for
+ * @param base pointer to first element to search
+ * @param num  number of elements
+ * @param size size of each element
+ * @param cmp  pointer to comparison function
+ */
+void * lv_bsearch(const void * key, const void * base, uint32_t num, uint32_t size, int32_t (* cmp)(const void * key, const void * elt))
```
Comment (puzrin) [inline lv_misc/lv_math.c]: Probably, it's time to create `lv_utils.c`/`lv_utils.h` files for assorted things. This function does not seems to be math-related.

PS. And `lv_math_num_to_str` from this file is not related to math too.

**Comment by manison** [comment]
The comment at the top of the source https://github.com/torvalds/linux/blob/master/lib/bsearch.c indicates the code is provided under the GPLv2 license which I believe is not acceptable in LittlevGL code base.

Wouldn't it be also nice if one could use C library implementation of `bsearch` where available (e.g. based on some #define in _lv_conf.h_)?

**Comment by puzrin** [comment]
> Wouldn't it be also nice if one could use C library implementation of bsearch where available (e.g. based on some #define in lv_conf.h)?

Looks like increase of complexity without added value. Probably worth add comment to fn src, why custom implementation done.

**Comment by manison** [comment]
> Looks like increase of complexity without added value. Probably worth add comment to fn src, why custom implementation done.

I wouldn't consider conditional compilation based on a single #define as _increase of complexity_, moreover when there are already plenty of similar cases throughout LittlevGL codebase (see `LV_xxx_CUSTOM_xxx`). But you got a point – the question is whether it is worth it. The library implementation of `bsearch` might be optimized for the target platform:

> Despite the name, neither C nor POSIX standards require this function to be implemented using binary search or make any complexity guarantees.

Since the font glyph lookup is on hot path, the performance gain of using optimized version might not be negligible, especially on embedded platforms. Or might not, I don't know :)

**Comment by puzrin** [comment]
> But you got a point – the question is whether it is worth it. The library implementation of bsearch might be optimized for the target platform:

Premature optimizations are evil :). Optimization exists != optimization required here.

**Comment by embeddedt** [comment]
@manison @beibean 
> The comment at the top of the source https://github.com/torvalds/linux/blob/master/lib/bsearch.c indicates the code is provided under the GPLv2 license which I believe is not acceptable in LittlevGL code base.

Correct. GPL code is completely incompatible with the MIT license and cannot be taken.

However, this looks compatible: http://www.beedub.com/Sprite093/src/lib/c/stdlib/bsearch.c

**Comment by manison** [comment]
> Correct. GPL code is completely incompatible with the MIT license and cannot be taken.

Maybe add mention in CONTRIBUTING.md?

**Comment by beibean** [comment]
> Correct. GPL code is completely incompatible with the MIT license and cannot be taken.
> However, this looks compatible: http://www.beedub.com/Sprite093/src/lib/c/stdlib/bsearch.c

I took a look to that new bsearch algorithm and basically does the same as the one implemented but in _another_ manner.

I will change it when I have a chance.

Sorry about not being careful about the license type, I have not worked much with these issues.

> Probably, it's time to create lv_utils.c/lv_utils.h files for assorted things. This function does not seems to be math-related. PS. And lv_math_num_to_str from this file is not related to math too.

I see it right as well.

**Comment by embeddedt** [comment]
@manison
> Maybe add mention in CONTRIBUTING.md?

Good idea. I've added some information about that.

**Comment by kisvegabor** [comment]
This issue was very instructive for me about licences.  

So in summary:
1. Let's create `lv_misc/lv_utils.c/h`
2. Use the MIT compatible `bsearch`
3. Not use define to select between library/custom implementation but use the custom.

Do you agree with that?

**Comment by beibean** [comment]
> Let's create lv_misc/lv_utils.c/h

I'm OK with that. Shall I move `lv_math_num_to_str` as well like @puzrin suggested? In that case, name should change as well.

**Comment by kisvegabor** [comment]
@beibean 
> I'm OK with that. Shall I move lv_math_num_to_str as well like @puzrin suggested? In that case, name should change as well.

Yes, please. (The `lv_math_` prefix should be changed to `lv_util_`.)

**Comment by kisvegabor** [comment]
Merged!

Thank you very much!
